### PR TITLE
fix: empty git-upload-pack given on remote.fetch()

### PR DESCRIPTION
**Pull request by greenpau**
More info: This change prevents an attempt to upload pack
when UploadPackRequest is empty.

Partial Resolution: #328

Signed-off-by: Paul Greenberg <greenpau@outlook.com>

### CODE REVIEW

**Comment by mcuadros** [comment]
can you cover it with some test?

**Comment by greenpau** [comment]
> can you cover it with some test?

@mcuadros , will do! 👍 will also need to do some adjustments to the code.

**Comment by greenpau** [comment]
@mcuadros , moved the check to the `UploadPack` function. The was introduced to `./plumbing/transport/internal/common/common.go` in #311.

Here, I add the same to the  `UploadPack` functions in the following files.

* `./plumbing/transport/http/upload_pack.go`
* `./plumbing/transport/file/server.go`

Created `TestFetchShallows` test, where provided depth 0.

Receiving the following error in the unrelated `repository_test.go`.

```
FAIL: repository_test.go:640: RepositorySuite.TestPlainOpenNotExistsDetectDotGit

repository_test.go:646:
    c.Assert(err, Equals, ErrRepositoryNotExists)
... obtained = nil
... expected *errors.errorString = &errors.errorString{s:"repository does not exist"} ("repository does not exist")
```

I am not 💯 sure about the `TestFetchShallows`. Please take a look and let me know what to adjust in the test. Many thanks!

**Comment by magnusbaeck** [comment]
This patch doesn't make a difference to me, which perhaps isn't surprising since I'm not making a shallow fetch. What's puzzling to me is that the wanted SHA-1 in wants really exists in the haves list so it's clear why isEmpty returns true. But shouldn't a commit in wants be available in the git, i.e. a checkout or `git show` operation would succeed? If I comment out the conditional in upload_pack.go that returns ErrEmptyUploadPackRequest I get an empty packfile and a ref that points to a non-existent object.

I apologize if you think I'm derailing your PR, maybe my problem is something entirely different (mysterious problems are often a case of PEBKAC), but it might also hint at a different root cause whereas the current state of this PR only addresses a particular corner case. Either way, I'd be happy to help debug the problem.

**Comment by greenpau** [comment]
> I apologize if you think I'm derailing your PR, maybe my problem is something entirely different (mysterious problems are often a case of PEBKAC), but it might also hint at a different root cause whereas the current state of this PR only addresses a particular corner case. Either way, I'd be happy to help debug the problem.

@magnusbaeck , this is valuable discussion! 👍 Please challenge, propose something new, etc.

**Comment by magnusbaeck** [comment]
The reason my wanted SHA-1 ended up in haves was that HEAD contained that SHA-1. My code looks like this:

```go
	err = worktree.Checkout(&git.CheckoutOptions{
		Hash:  plumbing.NewHash(commitID),
		Force: true,
	})
	if errors.Is(err, plumbing.ErrObjectNotFound) {
		refSpec := fmt.Sprintf("+%s:tmp", ref)
		err = gr.git.FetchContext(ctx, &git.FetchOptions{
			RefSpecs: []gitconfig.RefSpec{gitconfig.RefSpec(refSpec)},
			Tags:     git.NoTags,
		})
		...
```

It seems the failing checkout operation still ends up storing the wanted SHA-1 in HEAD, so when the subsequent fetch operation computes the haves set it picks up the SHA-1 from HEAD, resulting in an empty (wants − haves) set and hence ErrEmptyUploadPackRequest.

I'll double check to make sure the SHA-1 didn't end up in HEAD by other means, but if my analysis is correct I'll file an issue for the checkout behavior unless there already is one.

**Comment by magnusbaeck** [comment]
> I'll double check to make sure the SHA-1 didn't end up in HEAD by other means, but if my analysis is correct I'll file an issue for the checkout behavior unless there already is one.

For posterity my unrelated problem was filed as #457.

File: remote_test.go [hunk 1]
```diff
@@ -1261,6 +1261,22 @@ func (s *RemoteSuite) TestUpdateShallows(c *C) {
 	}
 }
 
+func (s *RemoteSuite) TestFetchShallows(c *C) {
+	r := NewRemote(memory.NewStorage(), &config.RemoteConfig{
+		URLs: []string{s.GetBasicLocalRepositoryURL()},
+	})
+
+	s.testFetch(c, r, &FetchOptions{
+		Depth: 0,
```
Comment (sarabala1979) [inline remote_test.go:1270]: Can you test it with non-zero value?

**Comment by l0nax** [comment]
@greenpau @mcuadros Is there any reason this PR cannot be merged?

**Comment by greenpau** [comment]
> Is there any reason this PR cannot be merged?

@l0nax , I cannot merge that. Not a maintainer of this package.

**Comment by pjbgf** [comment]
@greenpau do you mind rebasing your PR and also adding a test case to validate non-zero values for Depth?

**Comment by greenpau** [comment]
@pjbgf , I rebased. Unfortunately, it has been a while since I looked into this code. Could you add a test yourself?

**Comment by AriehSchneier** [comment]
This can be closed as it was fixed in #792
